### PR TITLE
Use string as IDs

### LINKED ISSUE #5: Make Node ID and Client ID string types
String IDs make integration of Mir easier, as Strings are more general than numeric IDs.

### PR DESCRIPTION
**Pull request by dnkolegov**
This PR fixes #5.

### CODE REVIEW

File: pkg/deploytest/deployment.go [hunk 1]
```diff
@@ -228,7 +228,7 @@ func localGrpcTransport(nodeIds []t.NodeID, ownId t.NodeID) *grpctransport.GrpcT
 	// Each test replica is on the local machine - 127.0.0.1
 	membership := make(map[t.NodeID]string, len(nodeIds))
 	for _, id := range nodeIds {
-		p, err := strconv.Atoi(string(id))
+		p, err := strconv.Atoi(id.Pb())
```
Comment (matejpavlovic) [inline pkg/deploytest/deployment.go:231]: Could you extend the comment documenting the function (line 223-224), and potentially other places if appropriate, that in the test deployment the node ID strings must be parseable to decimal numbers?

File: pkg/iss/iss.go [hunk 2]
```diff
@@ -696,7 +696,7 @@ func (iss *ISS) validateSBMessage(message *isspb.SBMessage, from t.NodeID) error
 	// TODO: This lookup is extremely inefficient, computing the membership set on each message validation.
 	//       Cache the output of membershipSet() throughout the epoch.
 	if _, ok := membershipSet(iss.config.Membership)[from]; !ok {
-		return fmt.Errorf("sender of SB message not in the membership: %d", from)
+		return fmt.Errorf("sender of SB message not in the membership: %s", from)
```
Comment (matejpavlovic) [inline pkg/iss/iss.go:699]: Let's use `%v` so we don't need to change this again in the unlikely case we use a different type for IDs in the future (or in a variant of Mir)

File: pkg/iss/iss.go [hunk 3]
```diff
@@ -920,7 +920,7 @@ func sequenceNumbers(start t.SeqNr, step t.SeqNr, length int) []t.SeqNr {
 
 // reqStrKey takes a request reference and transforms it to a string for using as a map key.
 func reqStrKey(reqRef *requestpb.RequestRef) string {
-	return fmt.Sprintf("%d-%d.%v", reqRef.ClientId, reqRef.ReqNo, reqRef.Digest)
+	return fmt.Sprintf("%s-%d.%v", reqRef.ClientId, reqRef.ReqNo, reqRef.Digest)
```
Comment (matejpavlovic) [inline pkg/iss/iss.go:923]: Let's use `%v` so we don't need to change this again in the unlikely case we use a different type for IDs in the future (or in a variant of Mir)

File: pkg/crypto/pseudo.go [hunk 4]
```diff
@@ -58,7 +59,7 @@ func NodePseudo(nodes []t.NodeID, clients []t.ClientID, ownID t.NodeID, seed int
 	if c == nil {
 		if err == nil {
 			// Own ID was not found and Crypto module instantiation was not even attempted.
-			return nil, fmt.Errorf("ownID (%d) not found among nodes", ownID)
+			return nil, fmt.Errorf("ownID (%s) not found among nodes", ownID)
```
Comment (matejpavlovic) [inline pkg/crypto/pseudo.go:62]: Let's use `%v` so we don't need to change this again in the unlikely case we use a different type for IDs in the future (or in a variant of Mir)

File: pkg/crypto/pseudo.go [hunk 5]
```diff
@@ -108,7 +109,7 @@ func ClientPseudo(nodes []t.NodeID, clients []t.ClientID, ownID t.ClientID, seed
 	if c == nil {
 		if err == nil {
 			// Own ID was not found and Crypto module instantiation was not even attempted.
-			return nil, fmt.Errorf("ownID (%d) not found among clients", ownID)
+			return nil, fmt.Errorf("ownID (%s) not found among clients", ownID)
```
Comment (matejpavlovic) [inline pkg/crypto/pseudo.go:112]: Let's use `%v` so we don't need to change this again in the unlikely case we use a different type for IDs in the future (or in a variant of Mir)

File: pkg/iss/bucketgroup.go [hunk 6]
```diff
@@ -59,7 +59,9 @@ func (buckets bucketGroup) Select(bucketIDs []int) bucketGroup {
 // Thus, the same request may map to some bucket in one group and to a different bucket in a different group,
 // even if the former bucket is part of the latter group.
 func (buckets bucketGroup) RequestBucket(reqRef *requestpb.RequestRef) *requestBucket {
-	bucketID := int(reqRef.ClientId+reqRef.ReqNo) % len(buckets) // If types change, this might need to be updated.
+	// TODO: do we need use clientID here?
```
Comment (matejpavlovic) [inline pkg/iss/bucketgroup.go:62]: I guess we don't. Unless all clients submit requests in lock step, I can't really imagine it causing any trouble.

File: workers.go [hunk 7]
```diff
@@ -369,23 +371,23 @@ func (n *Node) processReqStoreEvents(eventsIn *events.EventList) (*events.EventL
 
 			// Store request data.
 			if err := n.modules.RequestStore.PutRequest(storeEvent.RequestRef, storeEvent.Data); err != nil {
-				return nil, fmt.Errorf("cannot store request (c%dr%d) data: %w",
+				return nil, fmt.Errorf("cannot store request (c%sr%d) data: %w",
```
Comment (matejpavlovic) [inline workers.go:374]: Let's use `%v` so we don't need to change this again in the unlikely case we use a different type for IDs in the future (or in a variant of Mir)

File: workers.go [hunk 8]
```diff
@@ -369,23 +371,23 @@ func (n *Node) processReqStoreEvents(eventsIn *events.EventList) (*events.EventL
...
 					storeEvent.RequestRef.ClientId,
 					storeEvent.RequestRef.ReqNo,
 					err)
 			}
 
 			// Mark request as authenticated.
 			if err := n.modules.RequestStore.SetAuthenticated(storeEvent.RequestRef); err != nil {
-				return nil, fmt.Errorf("cannot mark request (c%dr%d) as authenticated: %w",
+				return nil, fmt.Errorf("cannot mark request (c%sr%d) as authenticated: %w",
```
Comment (matejpavlovic) [inline workers.go:382]: Let's use `%v` so we don't need to change this again in the unlikely case we use a different type for IDs in the future (or in a variant of Mir)

File: workers.go [hunk 9]
```diff
@@ -369,23 +371,23 @@ func (n *Node) processReqStoreEvents(eventsIn *events.EventList) (*events.EventL
...
 					storeEvent.RequestRef.ClientId,
 					storeEvent.RequestRef.ReqNo,
 					err)
 			}
 
 			// Store request authenticator.
 			if err := n.modules.RequestStore.PutAuthenticator(storeEvent.RequestRef, storeEvent.Authenticator); err != nil {
-				return nil, fmt.Errorf("cannot store authenticator (c%dr%d) of request: %w",
+				return nil, fmt.Errorf("cannot store authenticator (c%sr%d) of request: %w",
```
Comment (matejpavlovic) [inline workers.go:390]: Let's use `%v` so we don't need to change this again in the unlikely case we use a different type for IDs in the future (or in a variant of Mir)

**Review by matejpavlovic — COMMENTED** [review]
Nice. Just added very minor comments. Should be easy to fix (or even dismiss).

File: pkg/deploytest/deployment.go (hunk 1 above)
Comment (dnkolegov) [reply pkg/deploytest/deployment.go:231]: ok

File: pkg/crypto/pseudo.go (hunk 4 above)
Comment (dnkolegov) [reply pkg/crypto/pseudo.go:62]: ok

File: pkg/crypto/pseudo.go (hunk 5 above)
Comment (dnkolegov) [reply pkg/crypto/pseudo.go:112]: ok

File: pkg/iss/bucketgroup.go (hunk 6 above)
Comment (dnkolegov) [reply pkg/iss/bucketgroup.go:62]: ok

File: workers.go (hunk 7 above)
Comment (dnkolegov) [reply workers.go:374]: ok

File: workers.go (hunk 8 above)
Comment (dnkolegov) [reply workers.go:382]: ok

**Pull request merged**